libnavigation-base/src/main/java/com/mapbox/navigation/base/route/NavigationRoute.kt

```diff
@@ -0,0 +1,180 @@
+@file:JvmName("NavigationRouteEx")
+
+package com.mapbox.navigation.base.route
+
+import com.mapbox.api.directions.v5.models.DirectionsResponse
+import com.mapbox.api.directions.v5.models.DirectionsRoute
+import com.mapbox.api.directions.v5.models.DirectionsWaypoint
+import com.mapbox.api.directions.v5.models.LegStep
+import com.mapbox.api.directions.v5.models.RouteLeg
+import com.mapbox.api.directions.v5.models.RouteOptions
+import java.util.UUID
+
+/**
+ * Wraps a route object used across the Navigation SDK features.
+ *
+ * @param directionsResponse the original response that that returned this route object.
+ * @param routeIndex the index of the route that this wrapper tracks
+ * from the collection of routes returned in the original response.
+ * @param routeOptions options used to generate the [directionsResponse]
+ * @param id unique ID of this route entity used for matching it against additional metadata.
+ * Unique ID doesn't mean that the underlying [directionsRoute] is unique.
+ * Multiple [NavigationRoute]s can represent similar or equal [DirectionsRoute].
+ */
+class NavigationRoute(
+    val directionsResponse: DirectionsResponse,
+    val routeIndex: Int,
+    val routeOptions: RouteOptions,
+    val id: String = UUID.randomUUID().toString()
+) {
+
+    /**
+     * [DirectionsRoute] that this [NavigationRoute] represents.
+     */
+    val directionsRoute = directionsResponse.routes()[routeIndex]!!.toBuilder()
```

Comment:
This would effectively add another wrapper, and another step needed to expose a new property.

I'm considering `NavigationRoute#directionsRoute` as just a convenience field, it could as well be just an extension function.